OpenRA.Mods.RA/Activities/LayMines.cs

```diff
@@ -18,72 +20,93 @@
 
 namespace OpenRA.Mods.RA.Activities
 {
-	// assumes you have Minelayer on that unit
-	class LayMines : Activity
+	// Assumes you have Minelayer on that unit
+	public class LayMines : Activity
 	{
+		readonly Minelayer minelayer;
+		readonly MinelayerInfo info;
+		readonly IEnumerable<AmmoPool> ammoPools;
+
+		public LayMines(Actor self)
+		{
+			minelayer = self.TraitOrDefault<Minelayer>();
+			info = self.Info.Traits.Get<MinelayerInfo>();
+			ammoPools = self.TraitsImplementing<AmmoPool>();
+		}
+
 		public override Activity Tick(Actor self)
 		{
-			if (IsCanceled) return NextActivity;
+			if (IsCanceled)
+				return NextActivity;
 
 			var movement = self.Trait<IMove>();
-			var limitedAmmo = self.TraitOrDefault<LimitedAmmo>();
-			if (!limitedAmmo.HasAmmo())
-			{
-				var info = self.Info.Traits.Get<MinelayerInfo>();
-
-				// rearm & repair at fix, then back out here to refill the minefield some more
-				var buildings = info.RearmBuildings;
-				var rearmTarget = self.World.Actors.Where(a => self.Owner.Stances[a.Owner] == Stance.Ally
-					&& buildings.Contains(a.Info.Name))
-					.ClosestTo(self);
-
-				if (rearmTarget == null)
-					return new Wait(20);
-
-				return Util.SequenceActivities(
-					new MoveAdjacentTo(self, Target.FromActor(rearmTarget)),
-					movement.MoveTo(self.World.Map.CellContaining(rearmTarget.CenterPosition), rearmTarget),
-					new Rearm(self, info.RearmSound),
-					new Repair(rearmTarget),
-					this);
-			}
 
-			var ml = self.Trait<Minelayer>();
-			if (ml.Minefield.Contains(self.Location) && ShouldLayMine(self, self.Location))
+			if (ammoPools != null)
+				foreach (var pool in ammoPools)
+				{
+					if (pool.Info.Name == info.AmmoPool && !pool.HasAmmo())
```

Comment:
This `if`, `foreach`, `if` can be simplified to
`if (ammoPools != null && ammoPools.Any(pool => pool.Info.Name == info.AmmoPool && !pool.HasAmmo()))`